Dewikification - Create content app

ks129 opened the pull request:
- Created app that show content pages without wiki.
- All articles is stored to repository.
- Also included guide, that teach how to write a article.
- Created website embed for these pages for Discord
- Created categories system, ~~but this allow only 1 level of categories~~ now this allows unlimited subcategory levels.
- Fixes #383 

![Screenshot 2020-09-20 at 10 36 51](https://user-images.githubusercontent.com/45097959/93803018-eb15a600-fc4c-11ea-861a-9ccf4b3aab84.png)
![Screenshot 2020-09-20 at 10 37 07](https://user-images.githubusercontent.com/45097959/93803023-ec46d300-fc4c-11ea-9e85-f5a9ae5a081d.png)
![Screenshot 2020-09-20 at 10 37 15](https://user-images.githubusercontent.com/45097959/93803028-ee109680-fc4c-11ea-9e28-117f0bcdf590.png)
![Screenshot 2020-09-20 at 10 37 25](https://user-images.githubusercontent.com/45097959/93803029-ee109680-fc4c-11ea-8ef2-9e4b30046298.png)

ks129 committed:
Add markdown dependency to parse MD to HTML

ks129 committed:
Specify markdown package version

ks129 committed:
Install PyYAML for YAML files parsing

ks129 committed:
Create base Guides app

ks129 committed:
Include guides app to project

ks129 committed:
Create CSS file for all Guides app pages

ks129 committed:
Create HTML template for all guides (+ categories) view

ks129 committed:
Create HTML template for single guide view

ks129 committed:
Create HTML template for guides category view

ks129 committed:
Create all guides (+ categories) displaying view

ks129 committed:
Create view for displaying single guide

ks129 committed:
Create view for displaying guide categories

ks129 committed:
Add __init__.py to guides app views directory

ks129 committed:
Define guides app URLs

ks129 committed:
Add guides app URLs to home app URLs

ks129 committed:
Add codeblock parsing to Markdown parser for guide

ks129 committed:
Add syntax highlight support for guides

ks129 committed:
Add OpenGraph meta tags to guides for Discord embeds

ks129 committed:
Add example guide "How to Write a Guide"

ks129 committed:
Remove unnecessary debug path displaying for 404 on guide view

ks129 committed:
Add handling for invalid category for category guide

ks129 committed:
Add test data for guides

ks129 committed:
Add unit tests for guides app

ks129 committed:
Move guides and categories getting logic to utils for better testability

ks129 committed:
Apply testability changes to views tests

ks129 committed:
Create tests for get_category guides utility function

ks129 committed:
Create tests for get_categories guides utility function

ks129 committed:
Update `get_guides` function for better testability

ks129 committed:
Create new function `_get_base_path` for guides for better patching

ks129 committed:
Create tests for `get_guides`

ks129 committed:
Fix errors in unit tests that is based on recent changes

ks129 committed:
Create test for `_get_base_path`

ks129 committed:
Move base path to constant for guides utils unit tests

ks129 committed:
Create tests for `get_guide` function

ks129 committed:
Fix error that came in when moved path to constant

ks129 committed:
Use `yaml.safe_load` instead plain load to avoid warning

ks129 committed:
Add OG meta to category view

ks129 committed:
Add more information to how to write a guide guide

lemonsaurus commented on pydis_site/apps/guides/views/guides.py:
```diff
@@ -0,0 +1,38 @@
+            full_path = os.path.join(guides_path, name)
+            if os.path.isdir(full_path):
+                with open(os.path.join(full_path, "_info.yml")) as f:
+                    category = yaml.load(f.read())
```
This is pretty old fashioned. Please use [pathlib](https://docs.python.org/3/library/pathlib.html) instead.

lemonsaurus commented on pydis_site/apps/guides/views/guides.py:
```diff
@@ -0,0 +1,38 @@
+            elif os.path.isfile(full_path) and name.endswith(".md"):
+                md = Markdown(extensions=['meta'])
+                with open(full_path) as f:
+                    md.convert(f.read())
```
you can read files with pathlib too - something like `Path(full_path).read_text()` would make this whole context processor unnecessary.

lemonsaurus commented on pydis_site/apps/guides/views/guide.py:
```diff
@@ -0,0 +1,56 @@
+            path = os.path.join(settings.BASE_DIR, "pydis_site", "apps", "guides", "resources", "guides", f"{guide}.md")
+            category_name = None
+        else:
+            dir_path = os.path.join(settings.BASE_DIR, "pydis_site", "apps", "guides", "resources", "guides", category)
```
Basically I just want you to go through this whole pull request and use pathlib.

lemonsaurus commented on pydis_site/apps/guides/views/guide.py:
```diff
@@ -0,0 +1,56 @@
+            {
+                "guide": html,
+                "metadata": md.Meta,
+                "last_modified": datetime.fromtimestamp(os.path.getmtime(path)).strftime("%dth %B %Y"),
```
Hmm, interesting. I'm not sure this will work, but it's a fun idea. I'm worried that this will just show the last time we built the docker container rather than the actual mtime of the file, since the files don't persist through redeployments - but I'm not sure. Maybe I'm wrong.

We should verify whether this works.

lemonsaurus commented on pydis_site/apps/guides/resources/guides/how-to-write-a-guide.md:
```diff
@@ -0,0 +1,60 @@
+## [Getting Started](#getting-started){: id="getting-started" }
+First, you have to have a good idea, that match with PyDis theme. We can't accept guides like *How to bake a cake*,
+*How to lose weigth*. These doesn't match with PyDis theme and will be declined. Most of guides theme should be server and Python, but there can be some exceptions, when they are connected with PyDis.
+Best way to find out is your idea good is to discuss about it in #dev-core channel. There can other peoples give their opinion about your idea. Even better, open issue in site repository first, then PyDis staff can see it and approve/decline this idea.
```
Maybe you meant #dev-contrib?

lemonsaurus commented on pydis_site/apps/guides/resources/guides/how-to-write-a-guide.md:
```diff
@@ -0,0 +1,60 @@
+
+## [Getting Started](#getting-started){: id="getting-started" }
+First, you have to have a good idea, that match with PyDis theme. We can't accept guides like *How to bake a cake*,
+*How to lose weigth*. These doesn't match with PyDis theme and will be declined. Most of guides theme should be server and Python, but there can be some exceptions, when they are connected with PyDis.
```
I don't think this document has to explain any kind of policy, just focus it on the technical details.

lemonsaurus commented on pydis_site/apps/guides/resources/guides/how-to-write-a-guide.md:
```diff
@@ -0,0 +1,60 @@
+- **Contributors:** All who have contributed to this guide. One person per-line, and they **have to be at same level**. When you edit guide, add your name to there.
+- **Relevant Links and Values:** Links that will be shown at right side. Both key's values have to be at same level, just like for contributors field.
+
+## [Content](#content){: id="content" }
```
Why can't we do this programatically? It's stupid to have to write HTML for every single header.

lemonsaurus reviewed (dismissed):
Okay, this is a great start, and there's some pretty good code in here, but we need to figure out a few things.

@python-discord/core-developers, I want you guys to chime in on this with your opinions:
- The last time we discussed this system, I remember people raising concerns about the name "Guides". Should it be called Guides, or something more generic? If we should call it something else, what do we call it? Maybe something like Content?

- @ks129 How GitHub-Flavored Markdown compatible is this? The idea was that people were supposed to use the GitHub website to write guides. I think we need to ensure we support the whole (or as much as possible of the) GitHub-Flavored Markdown spec, and we need to rewrite the guide for how to write a guide to include information about writing the guide using our website. **The most important thing to point out is that they can use something like https://github.com/python-discord/site/new/guides-app/pydis_site/apps/guides/resources/guides to write, edit, preview and post a pull request**

- Is it possible to put an image in one of these guides? How would that look?

- Have you considered `markdown2` and the extras listed here: https://github.com/trentm/python-markdown2/wiki/Extras ? Some of these may be useful, especially in our pursuit to make this GHMD compatible.

- I'm not sure if we need contributors and last modified dates on the guides. It might be nice, but contributors has to be manually maintained, and I'm not convinced your last modified method will work in the context of docker containers. Perhaps it's better to just remove it. It feels like a wiki-feature that we're just clambering on to. @python-discord/core-developers do you guys have an opinion on this?

- I don't really love the html blended into these markdown templates. The whole point of this system was to get away from HTML. I'd prefer we just scrap this feature completely. If we need `## header` to have a certain id, let's just solve that automatically instead of requiring our users to write html inside curly brackets. I don't really want HTML to be possible for these guides, only plain markdown.

ks129 committed:
Change guides system to content system

As this system will be used for more than just guides,
I had to do some refactoring to match this system with plans.
Basically now there isn't guides, but articles instead.

ks129 committed:
Move from standard OS path actions to Pathlib

ks129 committed:
Migrate content system from Python-Markdown to markdown2

ks129 committed:
Remove last modified field from article

ks129 committed:
Implement custom icons for articles

ks129 committed:
Apply recent changes to article writing guide

ks129 committed:
Fix content app linting

lemonsaurus committed:
Merge branch 'dewikification' into guides-app

ks129 committed:
Rename content -> articles for visual part

ks129 committed:
Fix mess with rebase

ks129 committed:
Relock Pipfile.lock

ks129 committed:
Simplify relevant links generator empty values filtering

ks129 committed:
Add repository configuration to settings

ks129 committed:
Move repository configuration to environment variables

ks129 committed:
Add dateutil dependency because datetime's fromisoformat don't work

ks129 committed:
Implement fetching contributors and last modification date from GitHub

ks129 committed:
Add information about icons to articles writing guide

ks129 committed:
Rename last guides test to articles tests

ks129 committed:
Cover fetching article GitHub information with tests

jerbob commented on pydis_site/apps/content/views/articles.py:
```diff
@@ -0,0 +1,18 @@
+            request,
+            "content/articles.html",
+            {"content": get_articles(), "categories": get_categories()}
+        )
```
I'd use a `TemplateView` here and define `get_context_data`:

https://docs.djangoproject.com/en/3.1/topics/class-based-views/generic-display/#adding-extra-context

jerbob commented on pydis_site/templates/content/articles.html:
```diff
@@ -0,0 +1,53 @@
+        <div class="container">
+            <nav class="breadcrumb is-pulled-left" aria-label="breadcrumbs">
+                <ul>
+                    <li class="is-active"><a href="/articles">Articles</a></li>
```
Use the `{% url %}` tag here:
```suggestion
                    <li class="is-active"><a href="{% url 'articles:articles' %}">Articles</a></li>
```

ks129 committed:
Merge remote-tracking branch 'up/dewikification' into guides-app

ks129 committed:
Let Django generate URL for guides section of resources index

ks129 committed:
Remove unnecessary namespace from content app URLs include

ks129 committed:
Use Django URL generation for content app templates

ks129 committed:
Use just plain strings for site repo information

ks129 committed:
Use Django URL generation for category URL

ks129 committed:
Port views to TemplateView for content app

ks129 committed:
Update content app views tests

jerbob commented on pydis_site/apps/content/utils.py:
```diff
@@ -0,0 +1,117 @@
+
+def _get_base_path() -> Path:
+    """Have extra function for base path getting for testability."""
+    return Path(settings.BASE_DIR, "pydis_site", "apps", "content", "resources", "content")
```
I don't see why there's a method for this - wouldn't it be worth moving the `Path` object to a constant in settings, and patching `django.conf.settings.CONTENT_BASE_PATH`?

jerbob commented on pydis_site/apps/content/views/category.py:
```diff
@@ -0,0 +1,21 @@
+        context["category_info"] = get_category(self.kwargs["category"])
+        context["content"] = get_articles(self.kwargs["category"])
+        context["category_name"] = self.kwargs["category"]
+        return context
```
```suggestion
        category = self.kwargs["category"]
        context = super().get_context_data(**kwargs)
        context["category_info"] = get_category(category)
        context["content"] = get_articles(category)
        context["category_name"] = category
        return context
```

ks129 committed:
Add articles path to settings

ks129 committed:
Move articles writing guide to PyDis guides directory

ks129 committed:
Create view for showing articles and categories

ks129 committed:
Create categories and articles listing template

ks129 committed:
Update URLs to match new unlimited categories articles

ks129 committed:
Add testing subcategory and article into it

ks129 committed:
Change article.html to match with unlimited categories

ks129 committed:
Delete old article and category views

ks129 committed:
Update guides route in resources.html

ks129 committed:
Fix linting in resources view

ks129 committed:
Update template name of articles index

ks129 committed:
Update utils to match with new unlimited categories system

ks129 committed:
Change views __init__.py

ks129 committed:
Update tests to match with recent unlimited subcategories changes

ks129 committed:
Remove old templates

ks129 committed:
Merge branch 'dewikification' into guides-app

kosayoda committed:
Update highlight.js script and use one dark style.

Unifies the highlighting with the events app.

kosayoda committed:
Style <pre> tag backgrounds same as <code>.

This makes it consistent with the django-wiki version of the code
blocks, which looks neater.

kosayoda committed:
Remove Github metadata feature.

This feature is still under implementation debate, so it will be further
discussed in another issue and implemeneted in a future PR.

kosayoda committed:
Simplify relevant_links configuration in markdown.

kosayoda committed:
Rename `articles` to `pages`.

Articles was a good name, but we want an `articles` category in
the future. `/pages/guides/` and `/pages/articles/` are clearer in name
than `/articles/guides/` and `/articles/articles/`.

kosayoda committed:
Simplify content app.

Rather than having two views for the base page and all other pages, all
pages now use the same view.
The view context handler is simplified to take advantage of pathlib
features.
The markdown folder is now /content/resources/* rather than
/content/resources/content/*, as the latter is unnecessary nesting.

kosayoda committed:
Fix content app tests.

kosayoda committed:
Fix content columns when there is no sidebar.

If the `relevant_links` field is not provided in the markdown, do not
restrict contents to 2/3 columns.

kosayoda committed:
Rename `short_description` to `description`.

Given that there is no "long description", just `description` conveys
the same information with brevity.

kosayoda committed:
Propogate markdown language data to highlight.js

kosayoda committed:
Update guide to reflect changes in events app.

Wording is also improved.

MarkKoz commented on pydis_site/apps/content/resources/guides/pydis-guides/how-to-contribute-a-page.md:
```diff
@@ -0,0 +1,88 @@
+## Writing the Page
+Files representing pages are in `.md` (Markdown) format, with all-lowercase filenames and spaces replaced with `-` characters.
+
+Each page must include required metadata, and optionally additional metadata to modify appearance of the page.
```
```suggestion
Each page must include required metadata, and optionally additional metadata to modify the appearance of the page.
```

MarkKoz commented on pydis_site/apps/content/resources/guides/pydis-guides/how-to-contribute-a-page.md:
```diff
@@ -0,0 +1,88 @@
+    Using Git: https://pythondiscord.com/pages/contributing/working-with-git/
+---
+
+Pages, which include guides, articles and other static content are stored in markdown files in the `site` repository on Github.
```
Serial comma + comma at the end of the apositive.
```suggestion
Pages, which include guides, articles, and other static content, are stored in markdown files in the `site` repository on Github.
```

MarkKoz commented on pydis_site/apps/content/resources/guides/pydis-guides/how-to-contribute-a-page.md:
```diff
@@ -0,0 +1,88 @@
+    Using Git: https://pythondiscord.com/pages/contributing/working-with-git/
+---
+
+Pages, which include guides, articles and other static content...
```
For consistency.
```suggestion
Pages, which include guides, articles, and other static content, ...
```

MarkKoz commented on pydis_site/apps/content/resources/guides/pydis-guides/how-to-contribute-a-page.md:
```diff
@@ -0,0 +1,88 @@
+
+### Required Fields
+- **title:** Easily-readable title for your article.
+- **description:** Short, 1-2 line description that describes the page.
```
```suggestion
- **description:** Short, 1-2 line description of the page's content.
```

MarkKoz commented on pydis_site/apps/content/resources/guides/pydis-guides/how-to-contribute-a-page.md:
```diff
@@ -0,0 +1,88 @@
+
+## Extended Markdown
+
+Apart from standard markdown, certain additions are available:
```
```suggestion
Apart from standard Markdown, certain additions are available:
```

MarkKoz commented on pydis_site/apps/content/resources/guides/pydis-guides/how-to-contribute-a-page.md:
```diff
@@ -0,0 +1,88 @@
+| -------------- | ---------------------- |
+| An item        | Another item           |
+| Big item       | Small item             |
+
```
Aren't tables part of "standard" Markdown?

kosayoda replied:
[It's not](https://daringfireball.net/projects/markdown/syntax) unless I'm mistaken, it's just that most markdown implementations support it.

MarkKoz replied:
Okay fair enough.

MarkKoz commented on pydis_site/apps/content/resources/guides/pydis-guides/how-to-contribute-a-page.md:
````diff
@@ -0,0 +1,88 @@
+Syntax highlighting is provided by `highlight.js`.
+To activate syntax highlighting, put the language directly after the starting backticks.
+
+```python
+import os
+
+path = os.path.join("foo", "bar")
+```
````
This should be in a code block so that the raw Markdown is visible.

MarkKoz commented on pydis_site/apps/content/tests/test_utils.py:
```diff
@@ -0,0 +1,150 @@
+class TestGetPage(TestCase):
+    @override_settings(PAGES_PATH=BASE_PATH)
+    def test_get_root_page_success(self):
+        """Check does this return page HTML and metadata when root page exist."""
```
This phrasing is awkward (applies to all test docstrings). It's also redundant to write "check" because it's pretty obvious that tests check things.

kosayoda replied:
I'm inclined to remove the docstrings then, given the test method names provide enough information about the tests.

MarkKoz replied:
Yeah that may be the case for many tests. Use your discretion. I'd rather keep them if they add to the readability of the code even if the English is a bit awkward. That being said, I think it's not enforced in this project to use docstrings in tests and there are tests elsewhere without them.

MarkKoz commented on pydis_site/apps/content/resources/_info.yml:
```diff
@@ -0,0 +1,2 @@
+name: Pages
+description: Guides, articles and pages hosted on the site.
```
```suggestion
description: Guides, articles, and pages hosted on the site.
```

MarkKoz commented on pydis_site/templates/content/listing.html:
```diff
@@ -0,0 +1,59 @@
+    <section class="section">
+        <div class="container">
+            <div class="content">
+                <h1>{{ category_info.name|default:"Pages" }}</h1>
```
Everything from the start up to here is very similar to what is in the pages template. Can it be refactored out into a more generic template that both can use? The only differences are: no highlight.js, pages don't have a default title, and `page` & `category_info` have different interfaces for their attributes.

MarkKoz commented on pydis_site/templates/content/listing.html:
```diff
@@ -0,0 +1,59 @@
+                            <i class="{{ data.icon_class|default:"fab" }} {{ data.icon|default:"fa-python" }} is-size-3 is-black has-icon-padding" aria-hidden="true"></i>
+                        </span>
+                        <a href="{% url "content:page_category" location=path|add:page %}">
+                                <span class="is-size-4 has-text-weight-bold">{{ data.title }}</span>
```
This is indented too far.

MarkKoz commented on pydis_site/templates/content/listing.html:
```diff
@@ -0,0 +1,59 @@
+                            <i class="fas fa-folder is-size-3 is-black has-icon-padding" aria-hidden="true"></i>
+                        </span>
+
+
```
1 blank line is enough.

MarkKoz commented on pydis_site/static/css/content/page.css:
```diff
@@ -0,0 +1,16 @@
+     * backgrounds of <pre> tags but bulma does, resulting in a weird off-white
+     * border.
+     */
+    background-color: #282c34;
```
Would setting it to `unset` or `inherit` work?

kosayoda replied:
It would not, since that would result in a white `<pre>` background and a cramped `<code>` block.

![image](https://user-images.githubusercontent.com/41782385/112306191-82ca7e80-8cda-11eb-8966-e987b081bdec.png)

As a workaround to avoid hardcoding, I reduced the padding of `<pre>` and increased the `<code>` padding so it looks the same, but doesn't rely on specifying a background color.

MarkKoz commented on pydis_site/apps/events/views/page.py:
```diff
@@ -5,7 +5,7 @@
 
 
-class PageView(TemplateView):
+class PagesView(TemplateView):
```
This didn't have to be renamed. It's also inconsistent with the name `PageOrCategoryView`. It seems Django prefers singular names for views, but I didn't look into it deeply. Was it renamed cause the URL is plural?

kosayoda replied:
This was a mistake; I accidentally touched the `events` app here.

MarkKoz commented on pydis_site/apps/content/views/page_category.py:
```diff
@@ -0,0 +1,56 @@
+        return super().dispatch(request, *args, **kwargs)
+
+    def get_template_names(self) -> t.List[str]:
+        """Checks does this use page template or listing template."""
```
Awkward phrasing; similar to the test docstrings. How about "Check if the view uses..."?

MarkKoz commented on pydis_site/apps/content/utils.py:
```diff
@@ -0,0 +1,59 @@
+
+
+def get_categories(path: Path) -> Dict[str, Dict]:
+    """Get all categories information."""
```
```suggestion
    """Get information for all categories."""
```

MarkKoz commented on pydis_site/apps/content/utils.py:
```diff
@@ -0,0 +1,59 @@
+    pages = {}
+
+    for item in path.iterdir():
+        if item.is_file() and item.name.endswith(".md"):
```
`glob` could be used over `iterdir`. At the least, `.suffix` should be used over `name.endswith`.

MarkKoz commented on pydis_site/apps/content/views/page_category.py:
```diff
@@ -0,0 +1,56 @@
+                "path": str(location)
+            } for location in self.location.parents
+        ]
+        context["breadcrumb_items"] = reversed(breadcrumb_items)
```
I think this will work
```suggestion
        context["breadcrumb_items"] = [
            {
                "name": utils.get_category(settings.PAGES_PATH / location)["name"],
                "path": str(location)
            } for location in reversed(self.location.parents)
        ]
```

kosayoda replied:
yep!

MarkKoz commented on pydis_site/apps/content/utils.py:
```diff
@@ -0,0 +1,59 @@
+        ]
+    )
+
+    return {"page": str(html), "metadata": html.metadata}
```
`"content"` would be a better name. It's a bit strange to see `page.page` in the templates.

MarkKoz commented on pydis_site/apps/content/utils.py:
```diff
@@ -0,0 +1,59 @@
+
+
+def get_pages(path: Path) -> Dict[str, Dict]:
+    """Get all root or category page names and their metadata."""
```
The phrasing of the docstring is a bit confusing. If I understand correctly, it's mapping the names of pages in a category at `path` to their metadata. Perhaps `get_category_pages` is a more appropriate function name.

MarkKoz commented on pydis_site/apps/content/utils.py:
```diff
@@ -0,0 +1,59 @@
+
+def get_page(path: Path) -> Dict[str, Union[str, Dict]]:
+    """Get one specific page. When category is specified, get it from there."""
+    if not path.exists() or not path.is_file():
```
`is_file` already returns `False` is the path doesn't exist.

MarkKoz commented on pydis_site/apps/content/utils.py:
```diff
@@ -0,0 +1,59 @@
+
+def get_category(path: Path) -> Dict[str, str]:
+    """Load category information by name from _info.yml."""
+    if not path.exists() or not path.is_dir():
```
`is_dir` already returns `False` is the path doesn't exist.

MarkKoz commented on pydis_site/apps/content/utils.py:
```diff
@@ -0,0 +1,59 @@
+    if not path.exists() or not path.is_dir():
+        raise Http404("Category not found.")
+
+    return yaml.safe_load(path.joinpath("_info.yml").read_text())
```
Specify an encoding explicitly to be safe.

MarkKoz commented on pydis_site/apps/content/utils.py:
```diff
@@ -0,0 +1,59 @@
+
+    for name in path.iterdir():
+        if name.is_dir():
+            categories[name.name] = get_category(path.joinpath(name.name))
```
Isn't `name` already equivalent to `path.joinpath(name.name)`? Also, `name` is a poor choice of a name for a path variable.

kosayoda replied:
I agree.

MarkKoz commented on pydis_site/apps/content/utils.py:
```diff
@@ -0,0 +1,59 @@
+        raise Http404("Page not found.")
+
+    html = markdown(
+        path.read_text(),
```
Specify an encoding explicitly to be safe.

MarkKoz commented on pydis_site/apps/content/utils.py:
```diff
@@ -0,0 +1,59 @@
+
+
+def get_page(path: Path) -> Dict[str, Union[str, Dict]]:
+    """Get one specific page. When category is specified, get it from there."""
```
The second sentence is unnecessary. Why would someone assume otherwise? It just reads the file directly located at the given path. It doesn't even know or care whether it's in a category.

MarkKoz commented on pydis_site/apps/content/views/page_category.py:
```diff
@@ -0,0 +1,56 @@
+        if self.full_location.is_dir():
+            context["categories"] = utils.get_categories(self.full_location)
+            context["category_info"] = utils.get_category(self.full_location)
+            context["content"] = utils.get_pages(self.full_location)
```
"pages"  would be a better name than "content". At this point, "content" implies that it stores the HTML content of pages.

MarkKoz reviewed (changes requested):
I haven't looked at tests too much yet but might in another round.

It would be nice to (eventually) have tests that validate the Markdown metadata and YAML of all pages.

kosayoda committed:
Improve phrasing and wording of documentation.

kosayoda committed:
Undo change to events app.

Renaming `PageView` to `PagesView` was accidental and unnecessary.

kosayoda committed:
Refactor common HTML into a base template.

kosayoda committed:
Improve variable and key names for page utils.

kosayoda committed:
Simplify pathlib code and specify file encoding.

kosayoda committed:
Replace `markdown2` with `markdown` and `python-frontmatter`.

This allows us to properly escape codeblocks within markdown, permalink
to headers on a page, and decouples getting metadata from a file and
getting generated HTML from the Markdown content.

kosayoda committed:
Refactor content app tests.

The tests uses pyfakefs to simulate a fake filesystem that is reused
over the content app tests.
Test coverage for the app is brought to 100%.

MarkKoz commented on pydis_site/apps/content/resources/guides/pydis-guides/how-to-contribute-a-page.md:
````diff
@@ -0,0 +1,145 @@
+Pages, which include guides, articles, and other static content,...
+```
+
+You can learn more about Markdown metadata [here](https://github.com/trentm/python-markdown2/wiki/metadata).
````
Is it still appropriate to link to markdown2 docs?

kosayoda replied:
The information still holds, but now that we're using a separate library for frontmatter which uses `pyyaml` to build the metadata instead of whatever homegrown parser `markdown2` uses, I'll remove the line.

MarkKoz reviewed (approved):
Nice

kosayoda committed:
Remove guide reference to `markdown2`.

Now that we use a separate library for frontmatter (which uses pyyaml),
we can simplify the guide and remove the reference to the old markdown
library.
